### DISCUSSION TITLE
What are common causes of `SQLite error 14: unable to open database file - while executing PRAGMA journal_mode = WAL`?

### DISCUSSION THREAD

**Discussion by dawilliams-gpsw** [groue/GRDB.swift#1351] [Q&A] [answered]
Hello again!

We have some analytics around a database upgrade from our old DB to GRDB. One of the errors we see (in ONE of thousands of analytics) is the following:

```
SQLite error 14: unable to open database file - while executing `PRAGMA journal_mode = WAL`
```

I'm assuming that's happening as a result of this line in `DatabasePool.init(path:configuration:)`:
```
let journalMode = try String.fetchOne(db, sql: "PRAGMA journal_mode = WAL")
```

I have found a few discussions previously in this repo about similar SQLite errors related to multiple processes using the database, but in our case there is only one DBPool for the entirety of the app and it is only used in one process and one method call. 

We have also seen the low disk space error, which makes sense as a possibility:
```
SQLite error 13: database or disk is full - while executing `CREATE TABLE grdb_issue_102 (id INTEGER PRIMARY KEY)`
```

So my question is— what are some common causes of SQLite error 14? Is it possible that SQL was able to create the DB, but not the WAL file? In that case, would we still receive SQLite error 13 rather than 14? The database file is written to the application support directory, and since the user did not receive error 13, I'd imagine that their disk space was adequate. Could there be another cause?

Cheers!
Daniel

**Comment by groue** [comment, marked as answer]
Hello @dawilliams-gpsw,

You'll find information about `SQLITE_FULL` (13) in https://github.com/groue/GRDB.swift/issues/1251.

For `SQLITE_CANTOPEN` (14), I don't have much information. You'll find in this repo several issues about `SQLITE_CANTOPEN` thrown by read-only connections on a WAL database without a `-wal` file (it took quite a long time to understand what was going on) - but this does not precisely apply to you.

Maybe a search for `SQLITE_CANTOPEN` in the [SQLite forums](https://sqlite.org/forum/forummain) could help? Or a new question in this forum?

**Reply by dawilliams-gpsw** [reply]
@groue — I was able to dig a bit further into our analytics and noticed that the user had 52 bytes free on their phone. 😅 

I wouldn't have expected the `PRAGMA journal_mode = WAL` command to write anything to disk since it's a `fetch` call, but perhaps SQLite is doing something under the hood there that requires some additional disk space. I'm guessing you're just passing these errors along from SQLite, so I suppose it's not really a GRDB issue.

I appreciate the insight, thank you!
